k8sutils/poddisruption.go

```diff
@@ -0,0 +1,181 @@
+package k8sutils
+
+import (
+	"context"
+
+	"github.com/banzaicloud/k8s-objectmatcher/patch"
+	"github.com/go-logr/logr"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	redisv1beta1 "redis-operator/api/v1beta1"
+)
+
+// CreateRedisLeaderPodDisruptionBudget check and create a PodDisruptionBudget for Leaders
+func ReconcileRedisLeaderPodDisruptionBudget(cr *redisv1beta1.RedisCluster) error {
+	role := "leader"
+	pdbName := cr.ObjectMeta.Name + "-" + role
+	if cr.Spec.RedisLeader.PodDisruptionBudget != nil && cr.Spec.RedisLeader.PodDisruptionBudget.Enabled {
+		labels := getRedisLabels(cr.ObjectMeta.Name, "cluster", role)
+		pdbMeta := generateObjectMetaInformation(pdbName, cr.Namespace, labels, generateStatefulSetsAnots())
+		pdbDef := generatePodDisruptionBudgetDef(cr, role, pdbMeta, cr.Spec.RedisLeader.PodDisruptionBudget)
+		return CreateOrUpdatePodDisruptionBudget(pdbDef)
+	} else {
+		// Check if one exists, and delete it.
+		_, err := GetPodDisruptionBudget(cr.Namespace, pdbName)
+		if err == nil {
+			return deletePodDisruptionBudget(cr.Namespace, pdbName)
+		} else if err != nil && errors.IsNotFound(err) {
+			// Its ok if its not found, as we're deleting anyway
+			return nil
+		}
+		return err
+	}
+}
+
+// CreateRedisFollowerPodDisruptionBudget check and create a PodDisruptionBudget for Followers
+func ReconcileRedisFollowerPodDisruptionBudget(cr *redisv1beta1.RedisCluster) error {
+	role := "follower"
+	pdbName := cr.ObjectMeta.Name + "-" + role
+	if cr.Spec.RedisFollower.PodDisruptionBudget != nil && cr.Spec.RedisFollower.PodDisruptionBudget.Enabled {
+		labels := getRedisLabels(cr.ObjectMeta.Name, "cluster", role)
+		pdbMeta := generateObjectMetaInformation(pdbName, cr.Namespace, labels, generateStatefulSetsAnots())
+		pdbDef := generatePodDisruptionBudgetDef(cr, role, pdbMeta, cr.Spec.RedisFollower.PodDisruptionBudget)
+		return CreateOrUpdatePodDisruptionBudget(pdbDef)
+	} else {
+		// Check if one exists, and delete it.
+		_, err := GetPodDisruptionBudget(cr.Namespace, pdbName)
+		if err == nil {
+			return deletePodDisruptionBudget(cr.Namespace, pdbName)
+		} else if err != nil && errors.IsNotFound(err) {
+			// Its ok if its not found, as we're deleting anyway
+			return nil
+		}
+		return err
+	}
+}
```

Comment:
@davidcollom can we convert these two functions into a single one and pass the role to it as an argument. Because most things look the same to me.